### ISSUE TITLE
[Bug]: CHGNet forces and stresses may be inconsistent with energy derivatives due to detached three-body geometry

### ISSUE THREAD

**Issue by wakamiya0315** [materialyzeai/matgl#834]
**Version**

4.0.3

**Which OS(es) are you using?**

Linux

**What happened?**

In the PyG CHGNet implementation, `create_directed_line_graph()` is called inside `torch.no_grad()`. This also detaches the returned bond vectors and distances used in the three-body features, potentially excluding their contributions to forces and stresses.

[Relevant code in v4.0.3](https://github.com/materialyzeai/matgl/blob/v4.0.3/src/matgl/models/_chgnet.py#L347-L352)

[#791](https://github.com/materialyzeai/matgl/pull/791) appears to retain this behavior as part of the DGL-to-PyG parity-preserving port. The comparison below examines its effect on E/F/S accuracy.

I compared the **current MatGL v4.0.3 implementation** with a local variant that removes only the `torch.no_grad()` context around this call in `CHGNet.forward()` (`src/matgl/models/_chgnet.py`, lines 349–352):

```python
with torch.no_grad():
    (
        lg_edge_index, lg_bond_vec, lg_bond_dist,
        _lg_pbc_offset, lg_src_bond_sign,
    ) = create_directed_line_graph(
        edge_index, pbc_offset, bond_vec, bond_dist, self.three_body_cutoff
    )
```

Only gradient tracking for this call was changed. The pretrained weights and all other model code were unchanged; no retraining was performed.

Both conditions were evaluated on the complete official MatPES 2025.2 test files:

- [PBE test set](https://huggingface.co/datasets/materialyze/matpes/resolve/47d2cc020cf913b5a48a3480136a128dddc0a92c/MatPES-PBE-2025.2-test.jsonl): 21,660 structures.
- [r2SCAN test set](https://huggingface.co/datasets/materialyze/matpes/resolve/47d2cc020cf913b5a48a3480136a128dddc0a92c/MatPES-R2SCAN-2025.2-test.jsonl): 19,328 structures.

All results below were calculated on an **NVIDIA H100 MIG 3g.47gb using CUDA**, with TF32 disabled. Environment: PyTorch 2.13.0, PyG 2.8.0.post1, ASE 3.28.0.

| Model | `no_grad()` block | Energy MAE (meV/atom) | Force MAE (meV/Å) | Stress MAE (GPa) |
|---|---|---:|---:|---:|
| CHGNet-PES-MatPES-PBE-2025.2.10 | Present (current v4.0.3) | 24.678 | 144.342 | 0.6517 |
| CHGNet-PES-MatPES-PBE-2025.2.10 | Removed | 24.678 | 73.746 | 0.3505 |
| CHGNet-PES-MatPES-r2SCAN-2025.2.10 | Present (current v4.0.3) | 23.658 | 170.059 | 0.6794 |
| CHGNet-PES-MatPES-r2SCAN-2025.2.10 | Removed | 23.658 | 82.483 | 0.3343 |

Energy MAE is averaged per structure after dividing by its atom count. Force MAE uses all Cartesian components, and stress MAE uses all nine tensor components. Raw VASP stresses were converted from kbar to GPa with a factor of `-0.1`.

Removing the block reduced force/stress MAEs by approximately **46–52%**, while total energies differed by at most **6.1 × 10⁻⁵ eV per cell** between the two implementations.

For the PBE checkpoint on the perturbed Si cell in the snippet below, I also compared the returned forces with central finite differences of each variant's energy: `F_FD = -(E(R+h) - E(R-h)) / (2h)`. This check used **CUDA float32 and h = 0.002 Å** on the same H100.

| Implementation | Maximum absolute difference between returned and finite-difference forces (eV/Å) |
|---|---:|
| Current MatGL v4.0.3 | 0.145536 |
| Gradient tracking restored | 0.000268 |

Given these results, would it be appropriate to keep bond vectors and distances connected to positions and strain, while constructing only the graph connectivity without gradients?

**Code snippet**

```python
# Minimal CUDA comparison on a perturbed Si cell.
from unittest.mock import patch

import matgl
import matgl.models._chgnet as chgnet_impl
import numpy as np
import torch
from ase.build import bulk
from ase.calculators.fd import calculate_numerical_forces
from matgl.ext.ase import PESCalculator

torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.set_float32_matmul_precision("highest")

# Replace PBE with r2SCAN to check the other checkpoint.
potential = matgl.load_model(
    "materialyze/CHGNet-PES-MatPES-PBE-2025.2.10"
).to("cuda").eval()

atoms = bulk("Si", "diamond", a=5.43)
atoms.positions[0] += [0.035, -0.021, 0.017]
atoms.calc = PESCalculator(potential, stress_unit="eV/A3", use_voigt=True)


def check():
    atoms.calc.reset()
    energy = atoms.get_potential_energy()
    forces = atoms.get_forces()
    stress = atoms.get_stress()
    fd_forces = calculate_numerical_forces(atoms, eps=0.002)
    max_error = np.max(np.abs(forces - fd_forces))
    return energy, forces, stress, max_error


print("Current MatGL v4.0.3 E/F/S and FD error:", check())

# Restore gradients for the call wrapped in no_grad in CHGNet.forward().
create_line_graph = chgnet_impl.create_directed_line_graph
with patch.object(
    chgnet_impl,
    "create_directed_line_graph",
    torch.enable_grad()(create_line_graph),
):
    print("With geometry gradients E/F/S and FD error:", check())
```

**Log output**

No exception was raised. All 40,988 structures completed successfully under both conditions.

**Comment by shyuep** [comment]
@bowen-bd Can you take a look?

**Comment by bowen-bd** [comment]
Hi @wakamiya0315, thanks a lot for the report! This is addressed in https://github.com/materialyzeai/matgl/pull/835
